### PR TITLE
Feature/multi region openapi providers

### PR DESCRIPTION
**Pull request by dikhan**
## Proposed changes

Please add as many details as possible about the change here. Does this Pull Request resolve any open issue? If so, please make sure to link to that issue:
                                                              
Fixes: #84 

## Type of change

What type of change does your code introduce to the provider? Please put an `x` (w/o heading/trailing white spaces) 
in the boxes that apply:

- [ ] Bug-fix (change that fixes current functionality)
- [x] New feature (change that adds new functionality)

## Checklist

Please put an `x` (w/o heading/trailing white spaces) in the boxes that apply:

- [x] I have read and followed the [CONTRIBUTING](https://github.com/dikhan/terraform-provider-api/blob/master/.github/CONTRIBUTING.md) guidelines
- [x] I have made sure code compiles correctly
- [x] I have run 'make test' locally from the terraform_provider_api folder and no errors were found
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added/updated necessary documentation (if appropriate)

### CODE REVIEW

File: docs/how_to.md [hunk 1]
```diff
@@ -947,7 +947,79 @@ and the value provided by the user.
...
+  region = "dub"
+}
+
+
+## this resource will be managed in the default provider region, in this case rst (as it's the first element in the 'x-terraform-provider-regions' comma separated value) and API calls will be made against service.api.rst.hostname.com
+resource "provider_resource" "my_resource_rst" {
+  name = "resource in rst"
+}
+
+## this resource will be managed with the provider with alias dub, hence the region will be deb and API calls will be made against service.api.dub.hostname.com
```
Comment (pschuegr) [inline docs/how_to.md]: deb => dub

File: docs/how_to.md [hunk 2]
```diff
@@ -947,7 +947,79 @@ and the value provided by the user.
...
+
+In order to support multi-region configuration, the following extensions must be set with the right values:
+
+#### Extensions
+
+The following extensions can be used in the root level. Read the according extension section for more information
+
+Extension Name | Type | Description
+---|:---:|---
+[x-terraform-provider-multiregion-fqdn](#xTerraformProviderMultiregionFQDN) | string | Defines the host that should be used when managing the resources exposed. The value of this extension effectively overrides the global host configuration, making the OpenAPI Terraform provider client make the API calls against the host specified in this extension value instead of the global host configuration. The protocols (HTTP/HTTPS) and base path (if anything other than "/") used when performing the API calls will still come from the global configuration. The value must be parametrised following the expected format (regex: (\\S+)(\\$\\{(\\S+)\\})(\\S+)) where the ${region} section identifies the spot that will be replaced by the region value. E,g: service.api.${region}.hostname.com.
```
Comment (pschuegr) [inline docs/how_to.md]: remove escapes from regex

File: docs/how_to.md [hunk 3]
```diff
@@ -947,7 +947,79 @@ and the value provided by the user.
...
+x-terraform-provider-multiregion-fqdn: "service.api.${region}.hostname.com"
+````
+
+This extension must be present with the correct parametrised value in order for multi-region to be enabled.
+
+##### <a name="xTerraformProviderRegions">x-terraform-provider-regions</a>
+
+This extension defines the different regions supported by the service provider. The values will be used in the 'x-terraform-provider-multiregion-fqdn'
+value to build the final FQDN with the right region. The default value set in the terraform provider will be the first
+element in the command separated list, in the exaple below that will be 'rst':
```
Comment (pschuegr) [inline docs/how_to.md]: exaple => example

File: openapi/openapi_v2_backend_configuration.go [hunk 4]
```diff
@@ -34,6 +38,66 @@ func (o specV2BackendConfiguration) getHost() (string, error) {
...
+	for _, r := range allowedRegions {
+		if r == region {
+			return nil
+		}
+	}
+	return fmt.Errorf("region %s not matching allowed ones %+v", region, allowedRegions)
+}
+
+func (o specV2BackendConfiguration) getDefaultRegion() (string, error) {
+	isMultiRegion, _, regions, err := o.isMultiRegion()
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration.go]: if `getDefaultRegion` is always called after `isMultiRegion` maybe consider accepting regions as a parameter?

File: openapi/openapi_v2_backend_configuration.go [hunk 5]
```diff
@@ -34,6 +38,66 @@ func (o specV2BackendConfiguration) getHost() (string, error) {
...
+	return regions[0], nil
+}
+
+func (o specV2BackendConfiguration) isMultiRegion() (bool, string, []string, error) {
+	if host, exists := o.spec.Extensions.GetString(extTfProviderMultiRegionFQDN); exists {
+		isMultiRegion, _ := openapiutils.IsMultiRegionHost(host)
+		if !isMultiRegion {
+			return false, "", nil, fmt.Errorf("'%s' extension value provided not matching multiregion host format", extTfProviderMultiRegionFQDN)
+		}
+		regionsExtensionValue, regionsExtensionExists := o.spec.Extensions.GetString(extTfProviderRegions)
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration.go]: may want to throw an error if this exists by itself

File: openapi/openapi_v2_backend_configuration_test.go [hunk 6]
```diff
@@ -124,6 +124,302 @@ func TestGetHost(t *testing.T) {
...
+			SwaggerProps: spec.SwaggerProps{
+				Swagger: "2.0",
+				Host:    "www.some-backend.com",
+			},
+		}
+		openAPIDocumentURL := "www.domain.com"
+		specV2BackendConfiguration, _ := newOpenAPIBackendConfigurationV2(spec, openAPIDocumentURL)
+		Convey("When getHostByRegion method is called with a NON existing region", func() {
+			_, err := specV2BackendConfiguration.getHostByRegion("nonExisting")
+			Convey("Then the error returned should be nil", func() {
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration_test.go]: should not be nil? doesn't match assertion

File: openapi/openapi_v2_backend_configuration_test.go [hunk 7]
```diff
@@ -124,6 +124,302 @@ func TestGetHost(t *testing.T) {
...
+			_, err := specV2BackendConfiguration.getDefaultRegion()
+			Convey("Then the error returned should be nil", func() {
+				So(err, ShouldNotBeNil)
+			})
+			Convey("And the error should match the expected message", func() {
+				So(err.Error(), ShouldEqual, "failed to get default region value: service is not multi-region")
+			})
+		})
+	})
+	Convey("Given a specV2BackendConfiguration that is multi-region", t, func() {
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration_test.go]: mention that it doesn't have any regions as part of the description

File: openapi/openapi_v2_backend_configuration_test.go [hunk 8]
```diff
@@ -124,6 +124,302 @@ func TestGetHost(t *testing.T) {
...
+					extTfProviderMultiRegionFQDN: "www.${region}.some-backend.com",
+					extTfProviderRegions:         "",
+				},
+			},
+		}
+		openAPIDocumentURL := "www.domain.com"
+		specV2BackendConfiguration, _ := newOpenAPIBackendConfigurationV2(spec, openAPIDocumentURL)
+		Convey("When getDefaultRegion() method is called", func() {
+			_, err := specV2BackendConfiguration.getDefaultRegion()
+			Convey("Then the error returned should be nil", func() {
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration_test.go]: should not be nil

File: openapi/openapi_v2_backend_configuration_test.go [hunk 9]
```diff
@@ -124,6 +124,302 @@ func TestGetHost(t *testing.T) {
...
+			Convey("Then host should be the parametrised host", func() {
+				So(host, ShouldEqual, "www.${region}.some-backend.com")
+			})
+			Convey("Then regions should contain the right regions", func() {
+				So(regions, ShouldContain, "rst1")
+				So(regions, ShouldContain, "dub1")
+			})
+		})
+	})
+	Convey("Given a specV2BackendConfiguration that is multi-region (with 'x-terraform-provider-regions' extension values being comma separated with NO spaces)", t, func() {
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration_test.go]: This is a lot of code to test the whitespace regex

File: openapi/openapi_v2_backend_configuration_test.go [hunk 10]
```diff
@@ -124,6 +124,302 @@ func TestGetHost(t *testing.T) {
...
+			},
+		}
+		openAPIDocumentURL := "www.domain.com"
+		specV2BackendConfiguration, _ := newOpenAPIBackendConfigurationV2(spec, openAPIDocumentURL)
+		Convey("When isMultiRegion() method is called", func() {
+			isMultiRegion, _, _, err := specV2BackendConfiguration.isMultiRegion()
+			Convey("Then the error returned should be nil", func() {
+				So(err, ShouldBeNil)
+			})
+			Convey("Then it should be multi region", func() {
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration_test.go]: it should _not_ be multi region

File: openapi/openapi_v2_backend_configuration_test.go [hunk 11]
```diff
@@ -124,6 +124,302 @@ func TestGetHost(t *testing.T) {
...
+					extTfProviderMultiRegionFQDN: "www.some-backend.com",
+					extTfProviderRegions:         "rst1",
+				},
+			},
+		}
+		openAPIDocumentURL := "www.domain.com"
+		specV2BackendConfiguration, _ := newOpenAPIBackendConfigurationV2(spec, openAPIDocumentURL)
+		Convey("When isMultiRegion() method is called", func() {
+			isMultiRegion, _, _, err := specV2BackendConfiguration.isMultiRegion()
+			Convey("Then the error returned should be nil", func() {
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration_test.go]: same... too much c/p in tests?

File: openapi/openapi_v2_backend_configuration_test.go [hunk 12]
```diff
@@ -124,6 +124,302 @@ func TestGetHost(t *testing.T) {
...
+			Convey("And the error should match the expected message", func() {
+				So(err.Error(), ShouldEqual, "'x-terraform-provider-multiregion-fqdn' extension value provided not matching multiregion host format")
+			})
+			Convey("Then it should be multi region", func() {
+				So(isMultiRegion, ShouldBeFalse)
+			})
+		})
+	})
+
+	Convey("Given a specV2BackendConfiguration that is multi-region but the x-terraform-provider-multiregion-fqdn does not have a parametrised value", t, func() {
```
Comment (pschuegr) [inline openapi/openapi_v2_backend_configuration_test.go]: description should cover provider regions being empty

File: openapi/openapiutils/openapi_utils.go [hunk 13]
```diff
@@ -73,3 +74,26 @@ func GetSchemaDefinition(definitions map[string]spec.Schema, ref string) (*spec.
...
+		repStr := fmt.Sprintf("${1}%s$4", region)
+		return regex.ReplaceAllString(overrideHost, repStr), nil
+	}
+	return "", nil
+}
+
+// IsMultiRegionHost checks whether the override host is parametrised following the format expected
+func IsMultiRegionHost(overrideHost string) (bool, *regexp.Regexp) {
+	regex, err := regexp.Compile("(\\S+)(\\$\\{(\\S+)\\})(\\S+)")
+	if err != nil {
```
Comment (pschuegr) [inline openapi/openapiutils/openapi_utils.go]: probably don't need to catch this

**Comment by pschuegr** [comment]
Some comments on the tests (probably for another PR):

- there are a lot of tests where the description doesn't match the assertion
- this suggests to me that there is a lot of cut-n-paste going on when writing tests; it might be better to look at writing table-driven tests
- there are also tests where the final assertion is obscured by string formats, etc.  This is fine for regression testing, but my preference is usually to make the assertions explicit - it means that the tests are not as robust to input changes, but they are a lot easier to understand.
- if you find yourself writing a huge blob of test code to test a really simple condition, or catching errors that you know aren't going to be thrown it's worth thinking about whether there's a better way to deal with it (usually by restructuring the code/tests).

File: docs/how_to.md (hunk 1 above)
Comment (dikhan) [reply docs/how_to.md]: done

File: docs/how_to.md (hunk 2 above)
Comment (dikhan) [reply docs/how_to.md]: done

File: docs/how_to.md (hunk 3 above)
Comment (dikhan) [reply docs/how_to.md]: done

File: openapi/openapi_v2_backend_configuration.go (hunk 4 above)
Comment (dikhan) [reply openapi/openapi_v2_backend_configuration.go]: +1, done

File: openapi/openapi_v2_backend_configuration_test.go (hunk 6 above)
Comment (dikhan) [reply openapi/openapi_v2_backend_configuration_test.go]: done

File: openapi/openapi_v2_backend_configuration_test.go (hunk 7 above)
Comment (dikhan) [reply openapi/openapi_v2_backend_configuration_test.go]: done

File: openapi/openapi_v2_backend_configuration_test.go (hunk 9 above)
Comment (dikhan) [reply openapi/openapi_v2_backend_configuration_test.go]: agreed, will be tackling unit test clean up soon - different PR

File: openapi/openapi_v2_backend_configuration_test.go (hunk 10 above)
Comment (dikhan) [reply openapi/openapi_v2_backend_configuration_test.go]: done

File: openapi/openapi_v2_backend_configuration_test.go (hunk 11 above)
Comment (dikhan) [reply openapi/openapi_v2_backend_configuration_test.go]: +1

File: openapi/openapi_v2_backend_configuration_test.go (hunk 12 above)
Comment (dikhan) [reply openapi/openapi_v2_backend_configuration_test.go]: done

File: openapi/openapiutils/openapi_utils.go (hunk 13 above)
Comment (dikhan) [reply openapi/openapiutils/openapi_utils.go]: +1

**Comment by dikhan** [comment]
@pschuegr agreed 100% with the comments above. I have got plans to take a deep look into cleaning up the unit tests and reduce the boiler plate while ensuring readability is improved. table-driven tests seems like a good candidate to consider, will look into that.

File: openapi/openapi_v2_backend_configuration.go (hunk 5 above)
Comment (pschuegr) [reply openapi/openapi_v2_backend_configuration.go]: i think it's fine to leave it out, but there's a potential for slight confusion if somebody has the regions extension but not the multiregion host extension... from the perspective of somebody who doesn't know the code, i don't think either one seems more important than the other - why would it error if `A and !B` but not error if `!A and B`?

**Pull request merged**